### PR TITLE
Fix sdl crash on shutdown issues

### PR DESCRIPTION
**Pull request by OHerasym**
### PR Changes

- Remove unused code from utils
- Fix logger shutdown
- Fix catching signal SIGTERM
- Add class ThreadCounter for debugging multi thread issues

Related: [APPLINK-25734](https://adc.luxoft.com/jira/browse/APPLINK-25734) , [APPLINK-26736](https://adc.luxoft.com/jira/browse/APPLINK-26736)

Old pull [request](https://github.com/smartdevicelink/sdl_core/pull/729)

### CODE REVIEW

**Comment by OHerasym** [comment]
Please review: @anosach-luxoft , @LuxoftAKutsan , @orhan-mehmedov, @dtrunov

File: src/components/utils/src/logger_posix.cc [hunk 1]
```diff
@@ -108,6 +108,18 @@ void logger::Logger::Impl::DeinitLogger() {
   set_log_level(LogLevel::LL_TRACE);
   delete message_loop_thread_;
   message_loop_thread_ = NULL;
+
+  // 3rd-party logger deinitialization
+  log4cxx::LoggerPtr rootLogger = log4cxx::Logger::getRootLogger();
+  log4cxx::spi::LoggerRepositoryPtr repository =
+      rootLogger->getLoggerRepository();
+  log4cxx::LoggerList loggers = repository->getCurrentLoggers();
+  for (log4cxx::LoggerList::iterator i = loggers.begin(); i != loggers.end();
```
Comment (orhan-mehmedov) [inline src/components/utils/src/logger_posix.cc]: @OHerasym, seems that for_each also can be used here.

Comment (OHerasym) [inline src/components/utils/src/logger_posix.cc]: @orhan-mehmedov , ce41745 , Done

**Comment by LuxoftAKutsan** [comment]
Reviewed

File: src/components/utils/include/utils/log_message_loop_thread.h [hunk 2]
```diff
@@ -87,7 +87,9 @@ class LogMessageLoopThread : public LogMessageLoopThreadTemplate {
   LogMessageLoopThread()
       : LogMessageLoopThreadTemplate("Logger", new LogMessageHandler()) {}
 
-  ~LogMessageLoopThread() {}
+  ~LogMessageLoopThread() {
+    logger::Logger::FlushLogger();
```
Comment (anosach-luxoft) [inline src/components/utils/include/utils/log_message_loop_thread.h]: @OHerasym do you have requirement for that?

File: src/components/utils/include/utils/threads/thread_counter.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+#ifndef SRC_COMPONENTS_UTILS_INCLUDE_UTILS_THREADS_THREAD_COUNTER_H_
+#define SRC_COMPONENTS_UTILS_INCLUDE_UTILS_THREADS_THREAD_COUNTER_H_
+
+#include "utils/threads/thread.h"
+
+namespace threads {
+/**
+ * @brief The ThreadCounter class is helpful for debugging purpose.
+ * Helps to find and resolve multi threading issues.
```
Comment (anosach-luxoft) [inline src/components/utils/include/utils/threads/thread_counter.h]: @OHerasym please add explanation what actually it should count

File: src/components/utils/include/utils/threads/thread_counter.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  /**
+ * @brief Increment method increments thread counter
+ */
+  static void Increment();
+  /**
+   * @brief Decrement method decrements thread counter
+   */
+  static void Decrement();
+  /**
+   * @brief GetValue method gets thread counter
```
Comment (anosach-luxoft) [inline src/components/utils/include/utils/threads/thread_counter.h]: @OHerasym `GetValue method gets thread counter value`

File: src/components/utils/src/threads/thread_counter.cc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  DCHECK(ThreadCounter::GetValue() == 0);
+}
+
+ThreadCounter::~ThreadCounter() {
+  DCHECK(ThreadCounter::GetValue() == 0);
+}
+
+void ThreadCounter::Increment() {
+  sync_primitives::AutoLock auto_lock(counter_lock_);
+  thread_counter_++;
```
Comment (anosach-luxoft) [inline src/components/utils/src/threads/thread_counter.cc]: @OHerasym precrement

File: src/components/utils/src/threads/thread_posix.cc [hunk 6]
```diff
@@ -299,6 +306,9 @@ void Thread::StopUnsafe() {
 
 Thread::~Thread() {
   if (0 != handle_) {
+#ifdef THREAD_COUNT
```
Comment (anosach-luxoft) [inline src/components/utils/src/threads/thread_posix.cc]: @OHerasym please do it after thread had been destroyed (at the end of destructor)

File: src/components/utils/src/signals_posix.cc [hunk 7]
```diff
@@ -50,7 +50,9 @@ void SigHandler(int sig) {
       SDL_INFO("SIGTERM signal has been caught");
       break;
     case SIGSEGV:
+      SDL_FLUSH_LOGGER();
       SDL_INFO("SIGSEGV signal has been caught");
+      abort();
```
Comment (anosach-luxoft) [inline src/components/utils/src/signals_posix.cc]: @OHerasym what was wrong earlier?

File: src/components/utils/src/logger_posix.cc [hunk 8]
```diff
@@ -108,6 +108,16 @@ void logger::Logger::Impl::DeinitLogger() {
...
+       ++i) {
+    log4cxx::LoggerPtr logger = *i;
+    logger->removeAllAppenders();
+  }
+  rootLogger->removeAllAppenders();
```
Comment (anosach-luxoft) [inline src/components/utils/src/logger_posix.cc]: @OHerasym please explain why are you going to do that

File: src/components/utils/src/signals_posix.cc [hunk 9]
```diff
@@ -56,6 +56,7 @@ void SigHandler(int sig) {
       break;
     default:
       SDL_INFO("Unexpected signal has been caught");
+      exit(EXIT_FAILURE);
```
Comment (anosach-luxoft) [inline src/components/utils/src/signals_posix.cc]: @OHerasym why??

File: CMakeLists.txt [hunk 10]
```diff
@@ -64,6 +64,7 @@ option(ENABLE_SANITIZE "Sanitize tool" OFF)
 option(ENABLE_SECURITY "Security Ford protocol protection" ON)
 option(ENABLE_HMI_PTU_DECRYPTION "Policy table update parsed by hmi" ON)
 option(ENABLE_EXTENDED_POLICY "Turns extended flow which requires embedded system interaction" ON)
+option(ENABLE_THREAD_COUNT "Turns thread count to log file" OFF)
```
Comment (anosach-luxoft) [inline CMakeLists.txt]: @OHerasym it's not clear for me what does it mean `Turns thread count to log file`

File: src/components/utils/include/utils/log_message_loop_thread.h (hunk 2 above)
Comment (OHerasym) [inline src/components/utils/include/utils/log_message_loop_thread.h]: @anosach-luxoft , without Flushing we will have not all logs

File: src/components/utils/src/signals_posix.cc (hunk 7 above)
Comment (OHerasym) [inline src/components/utils/src/signals_posix.cc]: @anosach-luxoft , earlier SDL didn't shutdown when SIGSEGV was called (infinite loop)

File: src/components/utils/src/signals_posix.cc (hunk 9 above)
Comment (OHerasym) [inline src/components/utils/src/signals_posix.cc]: @anosach-luxoft , took this case from PASA

File: CMakeLists.txt (hunk 10 above)
Comment (OHerasym) [inline CMakeLists.txt]: @anosach-luxoft , e40d9da , Done

File: src/components/utils/src/logger_posix.cc (hunk 8 above)
Comment (OHerasym) [inline src/components/utils/src/logger_posix.cc]: @anosach-luxoft , this is correct shutdown for 3-rd party logger, in another case we will have core crash.

File: src/components/utils/src/threads/thread_posix.cc (hunk 6 above)
Comment (OHerasym) [inline src/components/utils/src/threads/thread_posix.cc]: @anosach-luxoft , e40d9da , Done

File: src/components/utils/src/threads/thread_counter.cc (hunk 5 above)
Comment (OHerasym) [inline src/components/utils/src/threads/thread_counter.cc]: @anosach-luxoft , e40d9da , Done

File: src/components/utils/include/utils/threads/thread_counter.h (hunk 4 above)
Comment (OHerasym) [inline src/components/utils/include/utils/threads/thread_counter.h]: @anosach-luxoft , e40d9da , Done

File: src/components/utils/include/utils/threads/thread_counter.h (hunk 3 above)
Comment (OHerasym) [inline src/components/utils/include/utils/threads/thread_counter.h]: @anosach-luxoft , e40d9da , Done

**Comment by OHerasym** [comment]
Please review: @AGritsevich

**Comment by AGritsevich** [comment]
@OHerasym , Reviewed.

**Comment by anosach-luxoft** [comment]
Reviewed

**Comment by dtrunov** [comment]
@OHerasym Reviewed

**Pull request merged**